File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/EnvShebangUsageRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.stream.Stream;

/**
 * Rule that requires executable scripts to use /usr/bin/env shebang.
 */
public enum EnvShebangUsageRule implements HarnessCheckRule {
    INSTANCE;

    @Override
    public String category() {
        return "envShebangUsage";
    }
    private static final String CATEGORY = "envShebangUsage";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final String expectedPrefix = catNode.get("parameters").get("expectedPrefix").asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return HarnessCheckHelper.extractPaths(catNode.get("parameters").get("directories")).stream()
                .flatMap(dir -> {
                    final Path dirPath = root.resolve(dir);
                    try {
                        return HarnessCheckHelper.safeFileOrWalk(root, dirPath).stream();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .filter(Files::isExecutable)
                .flatMap(file -> {
                    try {
                        return validateShebang(root, file, expectedPrefix, severity).stream();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }

    private List<Finding> validateShebang(Path root, Path file, String expectedPrefix, String severity) throws MojoExecutionException {
        final String text = HarnessCheckHelper.readFile(root, file);
        return text.startsWith("#!") && !text.startsWith(expectedPrefix)
                ? List.of(new Finding(severity, CATEGORY, "executable script should use /usr/bin/env shebang: " + root.relativize(file)))
                : List.of();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/EnvShebangUsageRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.HarnessPsiResults.Finding
import ai.harness.gradle.HarnessCheck
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.isExecutable
import kotlin.io.path.readText
import kotlin.io.path.relativeTo

/**
 * Rule that requires executable scripts under certain directories to use /usr/bin/env shebang.
 */
object EnvShebangUsageRule : HarnessCheckRule() {
    override val category: String = "envShebangUsage"
    override fun applies(manifest: JsonObject): Boolean =
        manifest[category]?.jsonObject?.get("enabled")?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true

    override fun validate(
        manifest: JsonObject,
        root: Path,
    ): Collection<Finding> {
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        val messagesObj = catObj?.get("messages")?.jsonObject
        return when {
            catObj == null || parametersObj == null || messagesObj == null -> {
                emptyList()
            }

            else -> {
                val directories = HarnessCheck.stringArrayFrom(parametersObj, "directories")
                val expectedPrefix = HarnessCheck.stringFrom(parametersObj, "expectedPrefix")
                directories
                    .filter { dirPath ->
                        (root / dirPath).isDirectory()
                    }.flatMap { dirPath ->
                        val dir = root / dirPath
                        val (files, _) = HarnessCheck.walkSafe(root, dir)
                        files
                            .filter { file -> file.isExecutable() }
                            .filter { file ->
                                val firstLine = file.readText().lineSequence().firstOrNull() ?: ""
                                firstLine.startsWith("#!") && !firstLine.startsWith(expectedPrefix)
                            }.map { file ->
                                Finding(
                                    HarnessCheck.severityOf(manifest, category),
                                    category,
                                    HarnessCheck.stringFrom(messagesObj, "default").takeIf { message ->
                                        message.isNotEmpty()
                                    }
                                        ?: "executable script should use /usr/bin/env shebang: ${file.relativeTo(
                                            root,
                                        )}",
                                )
                            }
                    }
            }
        }
    }

    override fun renderPsiFindings(
        findings: List<ai.harness.gradle.PsiFinding>,
        manifest: JsonObject,
    ): Collection<Finding> = emptyList()

    override fun findPsiFindings(
        file: Path,
        root: Path,
        psiFactory: org.jetbrains.kotlin.psi.KtPsiFactory?,
    ): List<ai.harness.gradle.PsiFinding> = emptyList()
}
